### PR TITLE
add diff command

### PR DESCRIPTION
**Pull request by epwalsh**
Adds a `diff`  command for comparing two arbitrary model checkpoints. This is analogous to `git diff`, except that instead of comparing lines between two files, we are comparing `(key, tensor)` pairs from the checkpoints' state dictionaries.

## Example

For example, run this to compare the pre-trained RoBERTa large weights to a version that has been fine-tuned on the SQuAD 2.0 task:

```bash
allennlp diff \
      https://huggingface.co/roberta-large/resolve/main/pytorch_model.bin \
      https://storage.googleapis.com/allennlp-public-models/transformer-qa-2020-10-03.tar.gz!weights.th \
      --strip-prefix-1 'roberta.' \
      --strip-prefix-2 '_text_field_embedder.token_embedder_tokens.transformer_model.'
```

image

The `△ = x.xxx` means a tensor was modified by this amount (this is 2-norm distance between the corresponding tensor from the first checkpoint and the one from the second).

## Technical details

The `diff` function uses a modified version of the default `git diff` algorithm: the [Myers algorithm](http://www.xmailserver.org/diff2.pdf). The only real modification between our algorithm and the original Myers algorithm is that in addition to the "keep", "insert", and "remove" operations, we also consider a "modify" operation. This operation corresponds to keeping a parameter with the same key and shape, but modifying its weights.

## Road map

- [ ] Automatically handle other input path or non-path types, like AllenNLP archives, HF model names, etc.

## Outstanding questions

- Is this useful?
- What can we change or add to the output to make this more helpful?
- What kinds of visualizations can we create on top of this?

### CODE REVIEW

**Comment by epwalsh** [comment]
Pinging @AkshitaB, @ArjunSubramonian, @jacob-morrison, and @dirkgr for feedback.

File: allennlp/commands/diff.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        )
+        subparser.add_argument(
+            "--strip-prefix-2",
+            type=str,
+            help="""A prefix to remove from all of the second checkpoint's keys.""",
+        )
+        return subparser
+
+
+def load_state_dict(
```
Comment (ArjunSubramonian) [inline allennlp/commands/diff.py:66]: Is this function not present anywhere else in the AllenNLP library? It might be better to import it than copy it here, to help with maintenance.

File: allennlp/commands/diff.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        termcolor.cprint(f"-{self.key}, shape = {self.shape}", "red")
+
+
+class Modify(NamedTuple):
+    key: str
+    shape: Tuple[int, ...]
+    distance: float
+
+    def display(self):
+        termcolor.cprint(f"!{self.key}, shape = {self.shape}, △ = {self.distance:.4f}", "yellow")
```
Comment (ArjunSubramonian) [inline allennlp/commands/diff.py:181]: I personally don't think that the L2 distance is too insightful, especially when all the tensors are so high-dimensional.

Comment (ArjunSubramonian) [reply allennlp/commands/diff.py:181]: That being said, I can't think of a better single metric to succinctly describe how different two tensors are. If we have such a metric, I don't personally think that sorting by that metric is helpful. I would much rather prefer to see the differing parameters in the order in which they are used during forward propagation.

Comment (ArjunSubramonian) [reply allennlp/commands/diff.py:181]: I think a heatmap of the element-wise differences of 2D parameters could be very useful. I don't know how to extend this well to more than 2 dimensions though, perhaps aggregate over the channel dimension?

**Review by ArjunSubramonian — COMMENTED** [review]
Left some comments! Overall, looks great :D

File: allennlp/commands/diff.py (hunk 1 above)
Comment (epwalsh) [reply allennlp/commands/diff.py:66]: Yes, good point. It should definitely be elsewhere.

File: allennlp/commands/diff.py (hunk 2 above)
Comment (epwalsh) [reply allennlp/commands/diff.py:181]: Maybe we flatten the tensors, then bin the weights into a manageable number of bins, then do something with those bins, like show a bar chart of the distance between bins of two tensors?

File: allennlp/commands/diff.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        subparser = parser.add_parser(
+            self.name,
+            description=description,
+            help=description,
+        )
+        subparser.set_defaults(func=_diff)
+        subparser.add_argument(
+            "checkpoint1",
+            type=str,
+            help="""The URL or path to the first PyTorch checkpoint file (e.g. '.pt' or '.bin').""",
```
Comment (dirkgr) [inline allennlp/commands/diff.py:46]: It would be nice if these could also point to model archives.

File: allennlp/commands/diff.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        An ordered dictionary of the state.
+    """
+    state = torch.load(path, map_location="cpu")
+    out: Dict[str, torch.Tensor] = OrderedDict()
+
+    if ignore is not None and not isinstance(ignore, list):
+        # If user accidentally passed in something that is not a list - like a string,
+        # which is easy to do - the user would be confused why the resulting state dict
+        # is empty.
+        raise ValueError("'ignore' parameter should be a list, you silly goose!")
```
Comment (dirkgr) [inline allennlp/commands/diff.py:106]: 😆 We might need a more professional error message.

File: allennlp/commands/diff.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+        if strip_prefix and key.startswith(strip_prefix):
+            strip_prefix_used = True
+            new_key = key[len(strip_prefix) :]
+            if not new_key:
+                raise ValueError("'strip_prefix' resulted in an empty string for a key")
+
+        out[new_key] = state[key]
+
+    if strip_prefix_used is False:
```
Comment (dirkgr) [inline allennlp/commands/diff.py:141]: We don't do `!strip_prefix_used` anymore? Did I miss a memo?

File: allennlp/commands/diff.py (hunk 2 above)
Comment (dirkgr) [reply allennlp/commands/diff.py:181]: You don't think that L2 distance is going to show up meaningfully when you want to validate, for example, that your gradual unfreezing schedule works? Maybe we can do better than L2, but I think it's a great start.

File: allennlp/commands/diff.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                out[i] = Modify(step.key, step.shape, dist)
+    return out
+
+
+def checkpoint_diff(
+    state_dict_a: Dict[str, torch.Tensor], state_dict_b: Dict[str, torch.Tensor]
+) -> List[Union[Keep, Insert, Remove, Modify]]:
+    """
+    Uses a modified version of the Myers diff algorithm to compute a representation
+    of the diff between two model state dictionaries.
```
Comment (dirkgr) [inline allennlp/commands/diff.py:210]: I don't know that much about Myers, but isn't that only necessary if the order matters? Does the order of entries in the `state_dict` matter? I thought that's just alphabetical?

**Review by dirkgr — CHANGES_REQUESTED** [review]
I think this is quite a good feature. I wish we could measure how much people use it.

File: allennlp/commands/diff.py (hunk 5 above)
Comment (epwalsh) [reply allennlp/commands/diff.py:141]: The difference here is that `strip_prefix_used` could be `None`.

File: allennlp/commands/diff.py (hunk 6 above)
Comment (epwalsh) [reply allennlp/commands/diff.py:210]: The order is meaningful. It is the order in which the corresponding modules were registered. Generally this is the order of data flow.

File: allennlp/commands/diff.py (hunk 4 above)
Comment (epwalsh) [reply allennlp/commands/diff.py:106]: 😞 
```suggestion
        raise ValueError("'ignore' parameter should be a list")
```

File: allennlp/commands/diff.py (hunk 2 above)
Comment (ArjunSubramonian) [reply allennlp/commands/diff.py:94]: It could be helpful. Just unsure of how L2 is affected by curse of dimensionality. Will it actually change that much?

Comment (ArjunSubramonian) [reply allennlp/commands/diff.py:94]: @epwalsh I like the bins idea a lot!!

Comment (epwalsh) [reply allennlp/commands/diff.py:94]: Looking over the example I gave, there is certainly a strong correlation with the number of elements in a modified parameter and the corresponding Euclidean distance, suggesting this is not the best metric to use.

That said, I'm *pretty sure* any other L*-based metric (even L∞) would suffer from the same correlation unless we add an additional size-based normalization term. I don't know what that term would look like exactly though.

Comment (epwalsh) [reply allennlp/commands/diff.py:94]: Any way, the exact metric to use should be a configurable option IMO

Comment (epwalsh) [reply allennlp/commands/diff.py:94]: Maybe just normalizing by $\sqrt{n}$ is good enough. So then we are really doing

$\sqrt{ \frac{1}{n} \sum_{i=0}^n (x_i - y_i)^2 }$

Or the square root of the mean squared "error". Is this meaningful?

File: allennlp/commands/diff.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    state_dict_b: Dict[str, torch.Tensor],
+) -> List[Union[Keep, Insert, Remove, Modify]]:
+    out = cast(List[Union[Keep, Insert, Remove, Modify]], history)
+    for i, step in enumerate(out):
+        if isinstance(step, Keep):
+            a_tensor = state_dict_a[step.key]
+            b_tensor = state_dict_b[step.key]
+            with torch.no_grad():
+                dist = torch.nn.functional.mse_loss(a_tensor, b_tensor).sqrt()
+            if dist != 0.0:
```
Comment (epwalsh) [inline allennlp/commands/diff.py:114]: Should we worry about loss of precision here? I.e. maybe we want to check that it's within some small threshold of 0, not exactly 0.

Comment (epwalsh) [reply allennlp/commands/diff.py:114]: The threshold could be a configurable parameter.

File: allennlp/commands/diff.py (hunk 2 above)
Comment (dirkgr) [reply allennlp/commands/diff.py:94]: Are y'all running plugins that render LaTex properly in GitHub?

Comment (epwalsh) [reply allennlp/commands/diff.py:94]: Human renderer plugin.

**Comment by epwalsh** [comment]
@dirkgr @ArjunSubramonian, this is ready for another review.

Now handles different input types: checkpoint files, archives, huggingface model IDs.

File: allennlp/commands/diff.py [hunk 8]
```diff
@@ -287,9 +227,29 @@ def one(idx):
...
+        elif ".tar.gz!" in checkpoint:
+            return cached_path(checkpoint, extract_archive=True)
+        else:
+            return cached_path(checkpoint)
+    elif checkpoint_type == "huggingface":
+        from transformers.file_utils import (
+            hf_bucket_url,
+            WEIGHTS_NAME,
+            cached_path as hf_cached_path,
+        )
```
Comment (dirkgr) [inline allennlp/commands/diff.py:243]: Can this go away when https://github.com/allenai/allennlp/pull/5052 is merged? If so, I'd rather wait. #5052 looks like it's almost ready for merging.

Comment (dirkgr) [reply allennlp/commands/diff.py:243]: In fact, if that is so, the whole notion of a checkpoint type can go away, right?

**Review by dirkgr — COMMENTED** [review]
Looks good. I'm just wondering if we can not have checkpoint types at all once we have native HF hub integration.

File: allennlp/commands/diff.py (hunk 7 above)
Comment (ArjunSubramonian) [reply allennlp/commands/diff.py:140]: This is a great point that I just thought of as well! I think we could maybe ensure better precision if we actually did (a_tensor != b_tensor).any(). But, I also like the idea of a configurable parameter. This way, if people train two models with a small difference in implementation, they can also use this tool to identify how similar the weights are with different thresholds epsilon.

File: tests/commands/diff_test.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        captured = capsys.readouterr()
+        assert (
+            _clean_output(captured.out)
+            == """
+ a, shape = (3,)
+-b, shape = (3,)
+-c, shape = (3,)
++b, shape = (4,)
++d, shape = (3,)
+!e, shape = (3,), difference = 0.5774
```
Comment (ArjunSubramonian) [inline tests/commands/diff_test.py:125]: Could you add a small comment that explains how you computed 0.5774?

File: allennlp/commands/diff.py (hunk 7 above)
Comment (epwalsh) [reply allennlp/commands/diff.py:140]: That's a great point

**Comment by epwalsh** [comment]
Yea @dirkgr, I'll hold off until the hub PR is merged

File: tests/commands/diff_test.py (hunk 9 above)
Comment (epwalsh) [reply tests/commands/diff_test.py]: Done.

File: allennlp/commands/diff.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+    scale: float,
+    threshold: float,
+) -> List[Union[Keep, Insert, Remove, Modify]]:
+    out = cast(List[Union[Keep, Insert, Remove, Modify]], history)
+    for i, step in enumerate(out):
+        if isinstance(step, Keep):
+            a_tensor = state_dict_a[step.key]
+            b_tensor = state_dict_b[step.key]
+            with torch.no_grad():
+                dist = (scale * torch.nn.functional.mse_loss(a_tensor, b_tensor).sqrt()).item()
```
Comment (ArjunSubramonian) [inline allennlp/commands/diff.py:158]: Just curious, is there a reason you opted for sqrt?

File: allennlp/commands/diff.py [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+            "--strip-prefix-1",
+            type=str,
+            help="""a prefix to remove from all of the 1st checkpoint's keys.""",
+        )
+        subparser.add_argument(
+            "--strip-prefix-2",
+            type=str,
+            help="""a prefix to remove from all of the 2nd checkpoint's keys.""",
+        )
+        subparser.add_argument(
```
Comment (ArjunSubramonian) [inline allennlp/commands/diff.py:89]: I don't completely understand the purpose of the scale arg, could you explain its necessity?

File: allennlp/commands/diff.py (hunk 10 above)
Comment (epwalsh) [reply allennlp/commands/diff.py:158]: Yea, two reasons:

1. It gives you another interpretation of the formula, which is L2 norm, normalized by the sqrt of the number of elements in the tensor, and
2. It seems to put the numbers on a scale that is easier to look at

File: allennlp/commands/diff.py (hunk 11 above)
Comment (epwalsh) [reply allennlp/commands/diff.py:89]: We only show a limited number of decimal places (4) in the output, so if you care about differences on the order of 1e-5 or smaller, you may want to increase the `--scale` parameter by an order of magnitude or more. Does that make sense? Should I include that in the help string?

**Comment by epwalsh** [comment]
I had to make a small change/fix to how we deal with `hf://` URLs to make it work with files like `roberta-large/pytorch_model.bin` that are not associated with a user or organization.

`cached_path("hf://roberta-large/pytorch_model.bin")` fails with a 404. Instead this proposes we add an extra "/" after "hf://" to avoid ambiguity, so you'd do `cached_path("hf:///roberta-large/pytorch_model.bin")`. Note this is only necessary when there is no user/org associated with the identifier. I'm not crazy about this solution, so open to better alternatives.

@LysandreJik, @dirkgr

**Comment by epwalsh** [comment]
On second thought, I *really* don't like that. Another option would be to check if the `url_or_filename` has one "/" and ends with any of the standard file names ("config.json", "pytorch_model.bin", etc.). But there would still be edge cases for other non-standard file names.

**Comment by dirkgr** [comment]
Is there any way we can make `hf://roberta-large/pytorch_model.bin` work? Why is that difficult?

**Comment by epwalsh** [comment]
> Is there any way we can make hf://roberta-large/pytorch_model.bin work? Why is that difficult?

Strictly speaking, `hf://roberta-large/pytorch_model.bin` is ambiguous. It could mean:
1. the file `pytorch_model.bin` in the `roberta-large` repository, which is the intent of course, but it could also mean
2. the repository `pytorch_model.bin` (repos are allowed to contain periods) owned by the user/organization called `roberta-large`.

**Comment by dirkgr** [comment]
We could try 2, and if it fails, fall back to 1.

**Comment by epwalsh** [comment]
> We could try 2, and if it fails, fall back to 1.

I think it would be better to try 1 first. If we try 2 first, this creates a vulnerability:

Say you have a production system that is downloading `hf://roberta-large/pytorch_model.bin` and getting the result from (1), as expected, because (2) doesn't exist. But then some joker creates a fake user/org account called "roberta-large" with a repo called "pytorch_model.bin". Then your system starts failing because it's downloading this junk repo (2), instead of (1).

**Comment by dirkgr** [comment]
Excellent point!

**Comment by LysandreJik** [comment]
That's a great find indeed, thank you for spotting it and offering a fix! 1 then 2 sounds good to me, is there any way I can help?

**Comment by epwalsh** [comment]
I moved the fix to this PR: https://github.com/allenai/allennlp/pull/5141
